### ISSUE TITLE
pathfs/loopback: slowly leaks memory at 1MB/minute

### ISSUE THREAD

**Issue by rfjakob** [hanwen/go-fuse#110]
Testcase is this script: https://github.com/rfjakob/gocryptfs/blob/master/tests/stress_tests/extractloop.bash . Run as `./extractloop.bash -loopback`. It basically extracts linux-3.0.tar.gz, deletes it, repeat.

RSS keeps growing at about 1MB/minute.

```
$ while true; do echo $(date) $(grep VmRSS /proc/15878/status) ; sleep 60; done
Di 12. Jul 08:30:05 CEST 2016 VmRSS: 97608 kB
Di 12. Jul 08:31:05 CEST 2016 VmRSS: 103604 kB
Di 12. Jul 08:32:05 CEST 2016 VmRSS: 105692 kB
Di 12. Jul 08:33:05 CEST 2016 VmRSS: 107420 kB
Di 12. Jul 08:34:05 CEST 2016 VmRSS: 107924 kB
Di 12. Jul 08:35:05 CEST 2016 VmRSS: 107924 kB
Di 12. Jul 08:36:05 CEST 2016 VmRSS: 108176 kB
Di 12. Jul 08:37:05 CEST 2016 VmRSS: 108176 kB
Di 12. Jul 08:38:05 CEST 2016 VmRSS: 108176 kB
Di 12. Jul 08:39:05 CEST 2016 VmRSS: 109444 kB
Di 12. Jul 08:40:05 CEST 2016 VmRSS: 110416 kB
Di 12. Jul 08:41:05 CEST 2016 VmRSS: 110624 kB
Di 12. Jul 08:42:05 CEST 2016 VmRSS: 110624 kB
Di 12. Jul 08:43:05 CEST 2016 VmRSS: 110624 kB
Di 12. Jul 08:44:05 CEST 2016 VmRSS: 110624 kB
Di 12. Jul 08:45:05 CEST 2016 VmRSS: 110976 kB
Di 12. Jul 08:46:05 CEST 2016 VmRSS: 110976 kB
Di 12. Jul 08:47:05 CEST 2016 VmRSS: 112084 kB
Di 12. Jul 08:48:06 CEST 2016 VmRSS: 112084 kB
Di 12. Jul 08:49:06 CEST 2016 VmRSS: 112084 kB
Di 12. Jul 08:50:06 CEST 2016 VmRSS: 124908 kB
Di 12. Jul 08:51:06 CEST 2016 VmRSS: 124908 kB
Di 12. Jul 08:52:06 CEST 2016 VmRSS: 124908 kB
```
Extractloop output should look like this:
```
$ ./extractloop.bash -loopback
Testing go-fuse loopback
Mounted!
[pid 15886] Starting loop
[pid 15887] Starting loop
[pid 15886] Iteration 1 done, 30 seconds
[pid 15887] Iteration 1 done, 31 seconds
[pid 15886] Iteration 2 done, 31 seconds
[pid 15887] Iteration 2 done, 30 seconds
[pid 15887] Iteration 3 done, 29 seconds
[pid 15886] Iteration 3 done, 30 seconds
[pid 15886] Iteration 4 done, 29 seconds
[pid 15887] Iteration 4 done, 31 seconds
[pid 15887] Iteration 5 done, 29 seconds
[pid 15886] Iteration 5 done, 32 seconds
[pid 15887] Iteration 6 done, 35 seconds
[pid 15886] Iteration 6 done, 35 seconds
[pid 15887] Iteration 7 done, 36 seconds
[pid 15886] Iteration 7 done, 34 seconds
[pid 15887] Iteration 8 done, 30 seconds
[pid 15886] Iteration 8 done, 32 seconds
```

**Comment by hanwen** [comment]
I did a  quikc look at a memory profile, but nothing stood out. You can look around using this trick here

https://github.com/hanwen/go-fuse/commit/97b8907272a20fa08cec5b0424cc5160b0afe263

can you see increasing memory use in the memory profiles?

**Comment by rfjakob** [comment]
Excellent idea! While trying to reproduce I hit another (small) issue. I have a patch ready, but it looks like I cannot push to gerrithub anymore. The error message does not make sense to me, 63999 is an old commit.

```
$ git push gerrithub HEAD:refs/for/master
Counting objects: 15, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (14/14), done.
Writing objects: 100% (15/15), 2.11 KiB | 0 bytes/s, done.
Total 15 (delta 11), reused 0 (delta 0)
remote: Resolving deltas: 100% (11/11)
remote: Processing changes: refs: 1, done    
remote: 
remote: ERROR:  In commit 639990cf387dd119837b2b8beb5fb6bc39c13caf
remote: ERROR:  committer email address hanwen@google.com
remote: ERROR:  does not match your user account.
remote: ERROR:
remote: ERROR:  The following addresses are currently registered:
remote: ERROR:    jakobunt@gmail.com
remote: ERROR:
remote: ERROR:  To register an email address, please visit:
remote: ERROR:  https://review.gerrithub.io/#/settings/contact
remote: 
remote: 
To ssh://rfjakob@review.gerrithub.io:29418/hanwen/go-fuse
 ! [remote rejected] HEAD -> refs/for/master (invalid committer)
error: failed to push some refs to 'ssh://rfjakob@review.gerrithub.io:29418/hanwen/go-fuse'
```

**Comment by hanwen** [comment]
I think you might be trying to push a commit by me that was committed to github, but didn't reach gerrithub.

It turns out gerrithub wants full write access to my gh account, which the folks at google don't like, so I've revoked gerrithub's access. (maybe I'll setup a gerrit on googlesource.com instead, not sure.)

**Comment by rfjakob** [comment]
I see, created an old-school pull request instead.

**Comment by rfjakob** [comment]
I let extractloop run for 24h and logged the RSS (see plot below). I have also triggered a memory profile every minute. Looking at the plot, the RSS seems to stabilize after 6 hours, which is good enough for me. 

If you still want to take a look, I have attached all the profiles, the RSS data and my loopback binary in [24h.zip](https://github.com/hanwen/go-fuse/files/367178/24h.zip).

![24h-rss](https://cloud.githubusercontent.com/assets/286847/16893697/62cb8984-4b40-11e6-8848-aee20457ee9c.png)
